### PR TITLE
Required tags

### LINKED ISSUE #12: Required Tag Association
#### As a user I want all images to have at least one tag.

Acceptance criteria:
- [x] I cannot save an image without adding at least one tag.
- [x] An error message is associated with the appropriate input field when a
  tag is not provided.

Dependencies:
- Image Tags

### PR DESCRIPTION
**Pull request by conproctor**
Closes #12

### CODE REVIEW

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -2,18 +2,18 @@
 
 class ImagesControllerTest < ActionController::TestCase
   test 'index' do
-    Image.create!(url: "https://imgur.com/1.jpg")
-    Image.create!(url: "https://imgur.com/2.jpg")
+    Image.create!(url: 'https://imgur.com/1.jpg', tag_list: 'required')
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: One fun trick to minimize needing to update many test cases whenever you add a required attribute is to use a private `image_create` function in your test that takes keyword arguments to provide a custom value. Something like:

```ruby
def image_create(url: 'http://imgur.com/1.jpg', tag_list: 'required')
  Image.create! url: url, tag_list: tag_list
```

Then you need only update that method to add a new required argument, and you can be clear within your tests via the omission of a keyword argument which arguments must be changed to be different than the default.

**Comment by bboe** [comment]
👍

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (conproctor) [inline test/controllers/images_controller_test.rb]: I like that pattern a lot, I'll keep it in mind for the future :+1:

Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: Agree with this completely, but I would prefer the method name as `create_image`

Comment (conproctor) [inline test/controllers/images_controller_test.rb]: Sightly related, when should one switch from a private `create` function to using a factory/exemplar?

Comment (bboe) [inline test/controllers/images_controller_test.rb]: > Agree with this completely, but I would prefer the method name as create_image

👍 

> when should one switch from a private create function to using a factory/exemplar?

If I recall correctly (I'm not the best with specific terminology) having a function to create an object is itself a factory. If by using a tool to help with factories, or a separate exemplar, I think that comes into play when you realize a need for having them.

In APM for instance there are dependent objects, where a unit may requires a property (separate object) to created. At this point it would be weird to have one test depend on another test class for code to prepare it's objects, and you certainly don't want to duplicate the same factories in different places. This is when it becomes useful. Also note that  the exemplar pattern is one we're falling out of favor for.
